DOC: reverse rolling window (#38627)

mdhsieh opened the pull request:
This adds the suggested changes from closed pull request (#38627):
https://github.com/pandas-dev/pandas/pull/39091/files/80819f65402e46caa29ff7192b34d43f91c1a0ee
After updated commit.
- remove noqa
- change line back to single quotes
- rename `df1` to `reversed_df`
- add reference to `FixedForwardWindowIndexer`
- remove repeated `df` initialization in code example
- rename `df2` to `reversed_df`
- move code example

- [x] closes #38627
- [ ] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

mdhsieh committed:
DOC: Add examples to do reversed rolling window (#38627).

File is copied from @vangorade last changes in pull request (#39091):
https://github.com/pandas-dev/pandas/pull/39091

Fixed typo to pass pre-commit codespell hook.

mdhsieh committed:
DOC: Add some suggested changes.

From pull request (#39091):
https://github.com/pandas-dev/pandas/pull/39091

- remove noqa
- change line back to single quotes
- rename df1 to reversed_df

mdhsieh committed:
DOC: Add reference to FixedForwardWindowIndexer.

mdhsieh committed:
DOC: Add more suggested changes.

From pull request (#39091):
https://github.com/pandas-dev/pandas/pull/39091

- remove repeated df initialization in code example
- rename df2 to reversed_df

mdhsieh committed:
DOC: Add suggested changes to reverse rolling window code examples (#38627).

Move code example as suggested in pull request (#39091):
https://github.com/pandas-dev/pandas/pull/39091

Compiled document to check it looked fine by running:
python make.py --single user_guide/window.rst

mzeitlin11 commented on doc/source/user_guide/window.rst:
```diff
@@ -50,16 +50,16 @@ As noted above, some operations support specifying a window based on a time offs
-   s = pd.Series(range(5), index=pd.date_range('2020-01-01', periods=5, freq='1D'))
-   s.rolling(window='2D').sum()
+   s = pd.Series(range(5), index=pd.date_range("2020-01-01", periods=5, freq="1D"))
+   s.rolling(window="2D").sum()
```
Looks like there are still some nonessential changes

mzeitlin11 commented on doc/source/user_guide/window.rst:
```diff
@@ -50,16 +50,16 @@ As noted above, some operations support specifying a window based on a time offs
-   df = pd.DataFrame({'A': ['a', 'b', 'a', 'b', 'a'], 'B': range(5)})
-   df.groupby('A').expanding().sum()
+   df = pd.DataFrame({"A": ["a", "b", "a", "b", "a"], "B": range(5)})
+   df.groupby("A").expanding().sum()
```
same here and other places

mdhsieh replied:
Ok, changed back

mzeitlin11 commented on doc/source/user_guide/window.rst:
```diff
@@ -381,8 +394,8 @@ two :class:`Series` or any combination of :class:`DataFrame`/:class:`Series` or
-  passed then computes the statistic for each pair of columns, returning a :class:`DataFrame` with a
-  :class:`MultiIndex` whose values are the dates in question (see :ref:`the next section
+  passed then computes the statistic for each pair of columns, returning a
+  ``MultiIndexed DataFrame`` whose ``index`` are the dates in question (see :ref:`the next section
```
This change looks suspect, think you might need to merge master if you didn't add this change

mzeitlin11 commented on doc/source/user_guide/window.rst:
```diff
@@ -592,7 +601,7 @@ The following formula is used to compute exponentially weighted mean with an inp
 
 .. math::
 
-    y_t = \frac{\sum_{i=0}^t 0.5^\frac{t_{t} - t_{i}}{\lambda} x_{t-i}}{\sum_{i=0}^t 0.5^\frac{t_{t} - t_{i}}{\lambda}},
```
same comment w respect to merging master as above

mdhsieh replied:
Ok merged

mzeitlin11 replied:
if you check the diff, looks unchanged. You might need to fetch first, see https://pandas.pydata.org/docs/development/contributing.html#working-with-the-code for a workflow on updating a pr

mdhsieh replied:
Okay, I added the changes from master branch file since fetching and merging didn't change those lines..

mzeitlin11 reviewed (commented):
Thanks for the pr @mdhsieh! Some initial comments

mdhsieh committed:
DOC: Revert more double quotes to single quotes.

Unecessary changes.

mdhsieh committed:
Merge remote-tracking branch 'upstream/master' into 38627-doc-reverse-rolling-window

mdhsieh committed:
DOC: Change binary window description to match master branch.

Description was not modified after fetching and merging, so pasted from master branch:
https://github.com/pandas-dev/pandas/blob/master/doc/source/user_guide/window.rst

mdhsieh committed:
DOC: Change exponentially weighted mean formula to one in master branch.

Since fetching and merging didn't modify this line.

mroeschke reviewed (changes requested):
Could you just add the new section and save cleanups for another PR (so just add content)?

It still appears that things changed that were not supposed to in the PR.

mdhsieh commented:
> Could you just add the new section and save cleanups for another PR (so just add content)?
> 
> It still appears that things changed that were not supposed to in the PR.

Yeah, I can do that.

mdhsieh committed:
DOC: Replace window.rst with file from pandas-dev master branch.

To only add reverse rolling window section and avoid any unwanted changes.

mdhsieh committed:
DOC: Add section with reverse rolling window example.

mdhsieh committed:
DOC: Add section markup.

Forgot to add.

mdhsieh commented:
Yeah I realized I accidentally deleted part of a code example.

```
For some problems...
.. ipython:: python

   from pandas.api.indexers import FixedForwardWindowIndexer
   indexer = FixedForwardWindowIndexer(window_size=2)
   df.rolling(indexer, min_periods=1).sum()
```

mdhsieh commented:
Updated, does this work now?

mroeschke commented on doc/source/user_guide/window.rst:
```diff
@@ -299,6 +299,38 @@ forward-looking rolling window, and we can use it as follows:
+   reversed_df = df[::-1].rolling("2s").sum()[::-1]
+   reversed_df
+
+Or we can also do it using :meth:`api.indexers.FixedForwardWindowIndexer` which basically creates window boundaries for fixed-length windows that include the current row.
```
This example is exactly the same as the one above this new section?

mdhsieh replied:
Right the `FixedForwardWindowIndexer` example is the same, my bad. Deleted it.

mroeschke commented on doc/source/user_guide/window.rst:
```diff
@@ -299,6 +299,38 @@ forward-looking rolling window, and we can use it as follows:
    indexer = FixedForwardWindowIndexer(window_size=2)
    df.rolling(indexer, min_periods=1).sum()
 
+Reverse rolling window
```
If this example has the same result as the `FixedForwardWindowIndexer` you added in this section, you can just continue in the section above and show that it's equivalent to the code block you added here.

mdhsieh replied:
Okay, sure.

mdhsieh committed:
DOC: Delete the same code example.

FixedForwaredWindowIndexer example is same as the one above the added section.

mdhsieh committed:
DOC: Remove section and first sentence.

Modify sentence slightly to show this example should output same result.

mroeschke commented on doc/source/user_guide/window.rst:
```diff
@@ -299,6 +299,24 @@ forward-looking rolling window, and we can use it as follows:
    indexer = FixedForwardWindowIndexer(window_size=2)
    df.rolling(indexer, min_periods=1).sum()
 
+We can also achieve this by using slicing in python, applying rolling aggregation, and then flipping the result as shown in example below:
```
Nit: Don't need `in python`

mdhsieh replied:
Sure, done.

mroeschke reviewed (commented):
One minor comment otherwise LGTM

mdhsieh committed:
DOC: Remove unecessary words.

Coding is already in Python.

mroeschke commented:
Thanks @mdhsieh. CI failure unrelated
